### PR TITLE
[Feature] Chip component

### PR DESCRIPTION
**Pull request by LJKaski**
## Description
This PR adds a Chip component to the library. Chip can be used as a button or simply as a styled way to present information. It is going to be used at least inside the upcoming Combobox component.

There are two variants of the chip. One is semantically a button, with an onClick prop and styles to present hover and active states. The other one is simply a styled span with no interactivity.

## Motivation and Context
Chip is a component needed for some features in the Suomi.fi services. It can also be used as a part of other components and modules such as Combobox.

## How Has This Been Tested?
So far tested by running locally and testing manually with firefox, chrome and edge.

## Screenshots (if appropriate):
image

## Release notes
Added Chip component

### CODE REVIEW

File: src/core/Chip/Chip.baseStyles.tsx [hunk 1]
```diff
@@ -0,0 +1,79 @@
+import { css } from 'styled-components';
+import { withSuomifiTheme, TokensAndTheme, SuomifiThemeProp } from '../theme';
+import { element, font, focus } from '../theme/reset';
+import { disabledCursor } from '../../components/utils/css';
+
+const removableStyles = ({ theme }: SuomifiThemeProp) => css`
+  &.fi-chip--removable {
```
Comment (aappoalander) [inline src/core/Chip/Chip.baseStyles.tsx:7]: Are these really only for removable chips?

File: src/core/Chip/Chip.baseStyles.tsx [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      border-radius: 16px;
+      }
+    }
+
+    &.fi-chip {
+      border-radius: 14px;
+      padding: ${theme.spacing.insetXxs} ${theme.spacing.insetL};
+      color: ${theme.colors.whiteBase};
+      background: ${theme.colors.highlightBase};
+      font: ${theme.typography.actionElementInnerText};
```
Comment (aappoalander) [inline src/core/Chip/Chip.baseStyles.tsx:43]: Is the bodyText above for something else or could we just use this one with the font function? And should the font function be used here as well? I think it applies e.g. word-breaks in addition to text design token.

File: src/core/Chip/Chip.baseStyles.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+      &:active {
+        background: ${theme.colors.highlightDark1};
+        }
+      }
+
+    & .fi-chip--icon{
+      height: 12px;
+      width: 12px;
+      transform: translateY(-0.35em);
```
Comment (aappoalander) [inline src/core/Chip/Chip.baseStyles.tsx:62]: Does this work well with browser zoom?

Comment (LJKaski) [reply src/core/Chip/Chip.baseStyles.tsx:62]: Tried it with the regular zoom function in Firefox and Chrome and it worked fine on both. Since we use pixel values for the fonts and elements, the text size setting doesn't really make a difference either way.

File: src/core/Chip/Chip.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    </Chip>
+  </div>
+  <div>
+    <Chip onClick={removeAction}>
+      Clickable chip without remove icon
+    </Chip>
+  </div>
+
+  <Chip.static>
+    Static chip with a long content that exceeds the size limit
```
Comment (aappoalander) [inline src/core/Chip/Chip.md:25]: This is a good example. Maybe even add the default max width to label? "exceeds the default maximum width of xx px." Since size limit is a bit ambiguous and could refer to character limit as well.

File: src/core/Chip/Chip.baseStyles.tsx (hunk 1 above)
Comment (LJKaski) [reply src/core/Chip/Chip.baseStyles.tsx:7]: Yes. When the icon is shown, the max width for the content has to be lower for the overflow styles and positioning to work correctly.

File: src/core/Chip/Chip.tsx [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+import { default as styled } from 'styled-components';
+import { baseStyles } from './Chip.baseStyles';
+import { HtmlButton, HtmlSpan } from '../../reset';
+import { TokensProp, InternalTokensProp } from 'core/theme';
+import { withSuomifiDefaultProps } from '../theme/utils';
+import { Icon } from '../Icon/Icon';
+import { logger } from '../../utils/logger';
+import { VisuallyHidden } from '../../components/Visually-hidden/Visually-hidden';
+
+const baseClassName = 'fi-chip';
```
Comment (aappoalander) [inline src/core/Chip/Chip.tsx:12]: We had this classnames object somewhere where you could just use e.g. classnames.disabled in the code. I personally preferred that over this more traditional solution.

File: src/core/Chip/Chip.tsx [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+const disabledClassName = `${baseClassName}--disabled`;
+const iconClassName = `${baseClassName}--icon`;
+const contentClassName = `${baseClassName}--content`;
+const removableClassName = `${baseClassName}--removable`;
+const buttonClassName = `${baseClassName}--button`;
+
+type ChipVariant = 'static' | 'default';
+
+interface InternalChipProps {
+  /** Chip element content */
```
Comment (aappoalander) [inline src/core/Chip/Chip.tsx:22]: should we add styles to allow overriding e.g. max-width without using the attached classnames? Not sure if we should, just wondering.

File: src/core/Chip/Chip.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
+```js
+import { Chip } from 'suomifi-ui-components';
+
+const removeAction = () => {
+  alert('Selection removed');
+};
+
+<>
+  <div styles={{ display: 'inline' }}>
```
Comment (aappoalander) [inline src/core/Chip/Chip.md:9]: This ends up in DOM as styles=[object, object]

Comment (LJKaski) [reply src/core/Chip/Chip.md:9]: Ah, it doesn't seem to be necessary after all. Just the div is enough 👍

File: src/core/Chip/Chip.tsx (hunk 6 above)
Comment (LJKaski) [reply src/core/Chip/Chip.tsx:22]: I think that this is a simple enough component that if you really want a chip that wide you can quite easily build your own or do some extra tricks with this one. I don't think it's worthwhile to support that out of the box.

File: src/core/Chip/Chip.tsx (hunk 5 above)
Comment (LJKaski) [reply src/core/Chip/Chip.tsx:12]: I think I used that in Checkbox. Didn't remember which one we agreed on. I'll implement it here as well.

File: src/core/Chip/Chip.test.tsx [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+  it('should match snapshot', () => {
+    const { container } = render(customClassChip);
+    expect(container).toMatchSnapshot();
+  });
+});
+
+describe('variants', () => {
+  const defaultChip = (
+    <Chip removable={true} onClick={() => null} actionLabel="Unselect">
```
Comment (aappoalander) [inline src/core/Chip/Chip.test.tsx:72]: Unselect sounds weird, should be deselect or remove?

File: src/core/Chip/Chip.tsx [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+    if (variant === 'static') {
+      return (
+        <HtmlSpan
+          className={classnames(baseClassName, className, {
+            [disabledClassName]: !!disabled,
+            [removableClassName]: !!removable,
+          })}
+          {...passProps}
+        >
+          <HtmlSpan className={contentClassName}>{children}</HtmlSpan>
```
Comment (aappoalander) [inline src/core/Chip/Chip.tsx:76]: Is this additional span necessary?

File: src/core/Chip/Chip.tsx [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+      <HtmlButton
+        className={classnames(baseClassName, buttonClassName, className, {
+          [disabledClassName]: !!disabled,
+          [removableClassName]: !!removable,
+        })}
+        disabled={disabled}
+        onClick={onClick}
+        {...passProps}
+      >
+        <HtmlSpan className={contentClassName}>{children}</HtmlSpan>
```
Comment (aappoalander) [inline src/core/Chip/Chip.tsx:90]: Same for this one, does the behaviour change if this span is removed and styles are applied directly to the parent element? This way content classname would not be needed either.

File: src/core/Chip/Chip.md [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    <Chip removable actionLabel="Unselect" onClick={removeAction}>
```
Comment (aappoalander) [inline src/core/Chip/Chip.md:10]: Unselect sounds weird, should be deselect or remove?

File: src/core/Chip/Chip.tsx (hunk 10 above)
Comment (LJKaski) [reply src/core/Chip/Chip.tsx:90]: I think it does. We need to be able to limit the size of the text content element in order to handle the overflow of the content correctly when an icon is present.

File: src/core/Chip/Chip.tsx (hunk 9 above)
Comment (LJKaski) [reply src/core/Chip/Chip.tsx:76]: This one might actually be redundant. I'll look into it.

**Comment by aappoalander** [comment]
Tests look good! Good work with those.

File: src/core/Chip/Chip.tsx (hunk 10 above)
Comment (LJKaski) [reply src/core/Chip/Chip.tsx:90]: We could maybe position the icon absolutely and handle this one without the span, but I'm not sure if that's a cleaner solution.

Comment (aappoalander) [reply src/core/Chip/Chip.tsx:90]: Sure, makes sense!

File: src/core/Chip/Chip.md (hunk 11 above)
Comment (LJKaski) [reply src/core/Chip/Chip.md:10]: I think deselect would be better. I'll change it.

File: src/core/Chip/Chip.md (hunk 4 above)
Comment (LJKaski) [reply src/core/Chip/Chip.md:25]: Then I need to rephrase the content, since the max width of the content depends on the variant while the max width of the component stays constant (290px). I'd be quite happy keeping this as it is.

File: src/core/Chip/Chip.baseStyles.tsx (hunk 2 above)
Comment (LJKaski) [reply src/core/Chip/Chip.baseStyles.tsx:43]: Just using actionElementInnerText in the main level font function seems to be enough, since there are no external labels or other things to consider here. Good point.

File: src/core/Chip/Chip.tsx (hunk 9 above)
Comment (LJKaski) [reply src/core/Chip/Chip.tsx:76]: Removing this would require some duplicate code and restructuring in the styles, so this might be a cleaner way to control the size and overflow.

File: src/core/Chip/Chip.md (hunk 4 above)
Comment (aappoalander) [reply src/core/Chip/Chip.md:25]: I would rephrase so that it says that max-width of the component is 290px. It's relevant information and helps devs when documented. Maybe something like "Static chip with long content that does not fit into components max-width of 290px."

File: src/core/Chip/Chip.tsx (hunk 5 above)
Comment (LJKaski) [reply src/core/Chip/Chip.tsx:12]: now changed.

File: src/core/Chip/Chip.baseStyles.tsx (hunk 2 above)
Comment (LJKaski) [reply src/core/Chip/Chip.baseStyles.tsx]: This is now fixed.

File: src/core/Chip/Chip.md (hunk 4 above)
Comment (LJKaski) [reply src/core/Chip/Chip.md]: Good suggestion. Put it into use.

File: src/core/Chip/Chip.md (hunk 7 above)
Comment (LJKaski) [reply src/core/Chip/Chip.md]: Should now be fixed.

File: src/core/Chip/Chip.md (hunk 11 above)
Comment (LJKaski) [reply src/core/Chip/Chip.md]: Now changed.

File: src/core/Chip/Chip.tsx (hunk 9 above)
Comment (aappoalander) [reply src/core/Chip/Chip.tsx]: Sounds reasonable, so let's keep it as it is even if not immediately obvious why it's there.

**Comment by aappoalander** [comment]
What about side margins? Looks kind of bad when chips are placed next to each other and there is no whitespace between them. Especially with the focus ring.

**Review by aappoalander — APPROVED** [review]

**Comment by aappoalander** [comment]
> What about side margins? Looks kind of bad when chips are placed next to each other and there is no whitespace between them. Especially with the focus ring.

As discussed in other forum, an example with margins is sufficient at this time and providing global margins would not be in line with other components.
